talisker/testing.py

```diff
@@ -21,17 +21,27 @@
 from __future__ import absolute_import
 
 from builtins import *  # noqa
-import sys
-import subprocess
+import logging
 import re
 import shlex
-import logging
+import subprocess
+import sys
+import tempfile
+import time
 
 import requests
 
 import talisker.logs
 
 
+if sys.version_info[0] == 2:
+    def temp_file():
+        return tempfile.NamedTemporaryFile('wb', bufsize=0)
+else:
+    def temp_file():
+        return tempfile.NamedTemporaryFile('wb', buffering=0)
```

Comment:
Would it perhaps be better to pass `bufsize=1` / `buffering=1`?  That should result in line buffering, which I think should mean that you don't have to flush the output file in `readline` (modulo my question below about whether that flush actually does anything, but if you're reading line-by-line then line buffering still seems appropriate).

Reply:
This is tricky to get right I think, so I was cautious, as buffering doesn't provide much benefit here - we're expecting <100 of log lines for a unit test case.

The py2/py3 version makes it interesting too.

In py3, the file-like object I pass to Popen has to be in binary mode, and thus cannot be line buffered, only block buffered, which we don't really want.

In py2, it only supports block buffering anyway, AFAICS, or -1 for the operating system default (which is block buffered for stdout, right?)

So having no buffering for output, and using readline() for input, seemed the best from a correctness pov.